### ISSUE TITLE
MediaSource's readyState is not 'open'

### ISSUE THREAD

**Issue by alexandercerutti** [google/shaka-player#2148]
Hi,
we are using Shaka Player in a very complex application, with also advertisings.
Since when we updated Shaka to latest `2.5.5` (from `2.4.5`), I begun to experience the following problem (formatted according our error reports format) for sure in Chrome for macOS. It is not a problem that keeps happening but one of those that happen sometimes and one of those that you cannot reproduce when you are trying to reproduce them. ( 😅)

The error is reported from `.load()` method's Promise catch:

```
{
    "error":{
        "stack": "Error: Failed to execute 'addSourceBuffer' on 'MediaSource': The MediaSource's readyState is not 'open'.
                  at http://localhost:3001/player-app.3edd5b3fe875f514028e.js:92710:282\n at Map.forEach (<anonymous>)
                  at Fa.b (http://localhost:3001/player-app.3edd5b3fe875f514028e.js:92710:135)\n at La (http://localhost:3001/player-app.3edd5b3fe875f514028e.js:92589:40)
                  at Ma.next (http://localhost:3001/player-app.3edd5b3fe875f514028e.js:92590:91)
                  at c (http://localhost:3001/player-app.3edd5b3fe875f514028e.js:92584:291)"
    },
    "message": "Failed to execute 'addSourceBuffer' on 'MediaSource': The MediaSource's readyState is not 'open'.",
    "code":"11"
}
```

Shaka player is bundled inside our application. For this reason the URLS are local.
Row `92710` is the following one (if needed) of the compiled js file, even if I don't know if Webpack compiles Shaka Player again.

```javascript
p.init = function(b, c) {
    var d = this;
    return t(function f() {
        var g;
        return z(f, function(f) {
            switch (f.j) {
                case 1:
                    return g = Cd, u(f, d.s, 2);
                case 2:
                    b.forEach(function(b, f) {
                        var h = oc(b.mimeType, b.codecs);
                        f == g.qa ? he(d, h) : (!c && MediaSource.isTypeSupported(h) || !Fd(h, f) || (d.l[f] = new Ed, h = Gd(f, h)), h = d.h.addSourceBuffer(h), d.i.w(h, "error", d.Tf.bind(d, f)), d.i.w(h, "updateend", d.ub.bind(d, f)), d.b[f] = h, d.c[f] = [])
                    }), v(f)
            }
        })
    })
};

function he(b, c) {
    b.a || (b.a = new Zd(b.g));
    be(b.a, c)
}
```

I'm not really sure how to solve it, but there are two things I know for sure:

1) This seemed to happen when we load the content after ads (we need to detach Shaka Player to allow our Ads provider to use the video tag) or when we change the content (like, from VOD to LIVE).
When we load the content we do this before loading the real URL:

```javascript
    if (this.player && !this.player.getMediaElement()) {
      await this.player.attach(this.videoNode, true);
    }
```

Where `this.player` is, obliviously, the Shaka Instance.

2) Recently, nothing changed in the flow of attaching our videoNode to shaka. The last change to that piece of code was made over 2 months ago (and we updated Shaka about 10 days ago).

Any clues about how to solve this and what we might be doing wrong?
Do anything important changed since `2.4.5` that might concern this issue?

Thank you very much!

Update 1:

I tried to print both `this.player` and `this.player.getMediaElement()` and both are available and set but the error keeps appearing.

**Comment by TheModMaker** [comment]
Are you using `player.detach()` to detach the Player?  Do you `await` the Promise returned from `detach`?  Maybe you are not calling `detach` in some path?  If external code changes the video element while Shaka Player is using it, things will break.  You could try printing the value of `video.src`.  It shouldn't change after an `attach`, so if you see it changing before the load, then something else is changing the video source and will break Shaka Player.

**Comment by alexandercerutti** [comment]
We are using `player.detach()`, but actually we are not using `await` on it. That's interesting.
Anyway, `detach()` is used in other pieces of code that do not matter in this case. I mean, I found that the problem happens when I change the content, from Live to Vod e vice versa (after the first time) and, especially after the Ads Prerolls. `detach()` is used before playing Midrolls and before playing Postrolls.

When we switch the content, we use `player.unload()` with a Promise catch in case of error.
Actually, it seems like we are not detaching Shaka before any Preroll. It might be worth looking at it.

Thank you for your suggestions, @TheModMaker. I'll look into them and check if that will still happen! 😄

**Comment by alexandercerutti** [comment]
@TheModMaker you were right!
I added the things we said above (`await`s and `detach`ing before prerolls) and it works again like a charm. 😄

Thank you very much!
